### PR TITLE
chore: [one click binding] Create a registry to maintain the list of the widget query generators

### LINKED ISSUE #21506: [task]: Datasource query generation registry
https://www.notion.so/appsmith/One-click-binding-dev-notes-b59349c84b0e404588555dbf3e291c0a?pvs=4#9dbba87720714deeae4e6b8324ad6f17

### PR DESCRIPTION
**Pull request by vsvamsi1**
## Description

Basic code to build datasource registry for one click binding.
Fixes #21506 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Jest

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/DataSourceAdaptors/MongoDB/index.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import BaseAdaptor from "../BaseAdaptor";
+
+export const MONGO_DB_PLUGIN_ID = "MongoDB";
```
Comment (mohanarpit) [inline app/client/src/DataSourceAdaptors/MongoDB/index.ts:3]: I'm assuming this is an early draft code. Either way, just a pointer to not hard code pluginIds in the client code base. This leads to a lot of problems in the future. If there's any configuration required for a particular plugin, please enhance & fetch this info from the API. This way, when we add a new datasource integration, we can integrate it seamlessly into this flow as well. 

Unfortunately, in the past, we've made this mistake with "Generate CRUD page" feature. That's why it's not easy to add more datasources to that feature. Let's not repeat that mistake here.

Comment (vsvamsi1) [reply app/client/src/DataSourceAdaptors/MongoDB/index.ts:3]: Sure @mohanarpit .... i feel i need to understand datasources flow in our application...do you know any of anyone in the team how is a SME of it or any documentation pertaining to this ....I want to understand the following terms
`"pluginType","pluginName", "pluginId"`

Comment (vsvamsi1) [reply app/client/src/DataSourceAdaptors/MongoDB/index.ts]: Hi @mohanarpit ....i have removed that segment of the code...

File: app/client/src/utils/WidgetQueryGeneratorRegistry.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { error } from "loglevel";
+
+export default class WidgetQueryGeneratorRegistry {
+  private static queryGeneratorMap = new Map();
+
+  static register(pluginId: string, queryGenerator: any) {
+    if (this.queryGeneratorMap.get(pluginId)) {
```
Comment (sbalaji1192) [inline app/client/src/utils/WidgetQueryGeneratorRegistry.ts:7]: use has method

File: app/client/src/utils/WidgetQueryGeneratorRegistry.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      error("Overwriting an query generator", pluginId);
```
Comment (sbalaji1192) [inline app/client/src/utils/WidgetQueryGeneratorRegistry.ts:8]: There is already a widget query generator with the given pluginID:

File: app/client/src/utils/WidgetQueryGeneratorRegistry.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+      return;
+    }
+
+    this.queryGeneratorMap.set(pluginId, queryGenerator);
+  }
+
+  static clear() {
+    WidgetQueryGeneratorRegistry.queryGeneratorMap = new Map();
```
Comment (sbalaji1192) [inline app/client/src/utils/WidgetQueryGeneratorRegistry.ts:17]: lets call map.clear method

File: app/client/src/utils/WidgetQueryGeneratorRegistry.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  static get(pluginId: string) {
+    const adaptor = this.queryGeneratorMap.get(pluginId);
```
Comment (sbalaji1192) [inline app/client/src/utils/WidgetQueryGeneratorRegistry.ts:21]: lets call the variable queryGenerator.

File: app/client/src/utils/WidgetQueryGeneratorRegistry.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    if (!adaptor) {
+      error("No query generator present", pluginId);
```
Comment (sbalaji1192) [inline app/client/src/utils/WidgetQueryGeneratorRegistry.ts:24]: Couldn't find the query generator with the given pluginId:

File: app/client/src/utils/WidgetQueryGeneratorRegistry.test.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import WidgetQueryGeneratorRegistry from "./WidgetQueryGeneratorRegistry";
+
+describe("WidgetQueryGeneratorRegistry", () => {
+  it("should be able to register a QueryGenerator", () => {
+    const somepluginId = "somePluginId";
+    const someQueryGenerator = {};
+
+    WidgetQueryGeneratorRegistry.register(somepluginId, someQueryGenerator);
+    expect(WidgetQueryGeneratorRegistry.get(somepluginId)).toBeTruthy();
```
Comment (sbalaji1192) [inline app/client/src/utils/WidgetQueryGeneratorRegistry.test.ts:9]: we need to assert that get function returns the same reference. not just truthy

File: app/client/src/utils/WidgetQueryGeneratorRegistry.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      return;
```
Comment (sbalaji1192) [inline app/client/src/utils/WidgetQueryGeneratorRegistry.ts:25]: return false here.

File: app/client/src/utils/WidgetQueryGeneratorRegistry.test.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  });
+
+  it("should return a falsey value when searching for an non existing generator", () => {
+    const nonExistingQueryGeneratopr = "someId";
+    expect(
+      WidgetQueryGeneratorRegistry.get(nonExistingQueryGeneratopr),
+    ).toBeFalsy();
+  });
+
+  afterAll(() => {
```
Comment (sbalaji1192) [inline app/client/src/utils/WidgetQueryGeneratorRegistry.test.ts:19]: we need to assertion for clear function as well..

**Review by sbalaji1192 — APPROVED** [review]

**Pull request merged**